### PR TITLE
Make Astlib's version type abstract

### PR DESCRIPTION
**Pull request by rgrinberg**
The first commit simplifies the interface of Astlib. In particular, there's an
`Astlib_intf` whose purpose is to avoid writing a subset of the signature of
`History` twice. It's possible to avoid this without an `Astlib_intf`, but it's
simpler to just copy a few signature.

The second commit introduces the abstraction version type. For now there's a
public `Astlib.Version.of_string`, but it's possible that in the future we'll
simply move all possible versions to `Astlib.Version` itself and this conversion
will be removed.

### CODE REVIEW

**Comment by ceastlund** [comment]
Why is it desirable to get rid of astlib_intf? Not only does it save us duplication, but it makes it clearer when reading history.mli which bindings are intended to be internal-only.

File: ast/cinaps/gen_traverse.ml [hunk 1]
```diff
@@ -530,7 +540,10 @@ let print_traverse_ml () =
 
 let print_traverse_mli () =
   Print.newline ();
-  let grammars = Astlib.History.versioned_grammars Astlib.history in
+  let grammars =
+    Astlib.History.versioned_grammars Astlib.history
+    |> List.map ~f:(fun (v, grammar) -> (Astlib.Version.to_string v, grammar))
```
Comment (ceastlund) [inline ast/cinaps/gen_traverse.ml:545]: Let's not map this over all our alists, if we're going to use a Version.t type, let's use it consistently, and only convert to string in the instances where we need to. I think it'll be clearer that way all around.

**Comment by rgrinberg** [comment]
> Why is it desirable to get rid of astlib_intf? Not only does it save us duplication, but it makes it clearer when reading history.mli which bindings are intended to be internal-only.

Well we can address these problems with a module that has a narrower scope: `history_intf.ml`:

```ocaml
module type Public :
   ...
end
```

`history.mli`:
```ocaml
include History_intf.Public

val ..
```

`astlib.mli`:

```
...
module History : History_intf.Public
...
```

There's no need to introduce an interface for the whole library.

I find an that an intermediate interface makes things hard to follow and usually produces worse documentation with our current tools (odoc, ocamldoc).

I don't feel so strongly about this, so we can revert it. But I'm fairly certain it will be brought up again once astlib is upstreamed.

**Comment by ceastlund** [comment]
Should be resolved by #63, which I just merged.
